### PR TITLE
fix update region size

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Make sure `end_key` in `influenced_regions` are in incremental order, and enhance test to cover it.

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -1795,14 +1795,14 @@ impl<T: Transport, C: PdClient> Store<T, C> {
```
Comment (huachaohuang) [inline src/raftstore/store/store.rs]: Can we test this function alone?

**Comment by siddontang** [comment]
Please add a description.

File: src/raftstore/store/store.rs (hunk 1 above)
Comment (siddontang) [reply src/raftstore/store/store.rs]: +1

**Comment by huachaohuang** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
PTAL

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -2661,3 +2634,99 @@ pub fn new_compaction_listener(ch: SendCh<Msg>) -> CompactionListener {
...
+
+fn calc_region_declined_bytes(
+    event: CompactedEvent,
+    region_ranges: &BTreeMap<Key, u64>,
+    bytes_threshold: u64,
+) -> Vec<(u64, u64)> {
+    // Calculate influenced regions.
+    let mut influenced_regions = vec![];
+    for (end_key, region_id) in
+        region_ranges.range((Excluded(event.start_key), Included(event.end_key.clone())))
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: why clone the end key but not clone start key here?

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -2661,3 +2634,99 @@ pub fn new_compaction_listener(ch: SendCh<Msg>) -> CompactionListener {
...
+    for (region_id, end_key) in influenced_regions.drain(..) {
+        let mut old_size = 0;
+        for prop in &event.input_props {
+            old_size += prop.get_approximate_size_in_range(&last_end_key, &end_key);
+        }
+        let mut new_size = 0;
+        for prop in &event.output_props {
+            new_size += prop.get_approximate_size_in_range(&last_end_key, &end_key);
+        }
+        last_end_key = end_key.clone();
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: no need clone here?

File: src/raftstore/store/store.rs (hunk 2 above)
Comment (zhangjinpeng1987) [reply src/raftstore/store/store.rs]: `end_key` will be used again.

**Comment by zhangjinpeng1987** [comment]
@siddontang PTAL

**Comment by zhangjinpeng1987** [comment]
@UncP

**Comment by UncP** [comment]
LGTM

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -2661,3 +2634,99 @@ pub fn new_compaction_listener(ch: SendCh<Msg>) -> CompactionListener {
...
+        .next()
+    {
+        influenced_regions.push((region_id, end_key.clone()));
+    }
+
+    // Calculate declined bytes for each region.
+    // `end_key` in influenced_regions are in incremental order.
+    let mut region_declined_bytes = vec![];
+    let mut last_end_key: Vec<u8> = vec![];
+    for (region_id, end_key) in influenced_regions.drain(..) {
```
Comment (overvenus) [inline src/raftstore/store/store.rs]: It looks like we don't need to `drain`.

File: src/util/rocksdb/event_listener.rs [hunk 5]
```diff
@@ -117,8 +117,8 @@ impl rocksdb::EventListener for CompactionListener {
             }
         }
 
-        let mut input_files = HashSet::default();
-        let mut output_files = HashSet::default();
+        let mut input_files = HashSet::with_capacity(info.input_file_count());
```
Comment (overvenus) [inline src/util/rocksdb/event_listener.rs]: How about `HashSet::with_capacity_and_hasher`?

Comment (zhangjinpeng1987) [reply src/util/rocksdb/event_listener.rs]: The numbers of output_files and input_files are not too large, there is no performance issue.

File: src/raftstore/store/store.rs (hunk 4 above)
Comment (zhangjinpeng1987) [reply src/raftstore/store/store.rs]: We need to take away end_key from it.

Comment (overvenus) [reply src/raftstore/store/store.rs]: So `for (region_id, end_key) in influenced_regions { .. }`, it is simple and neat.

Comment (huachaohuang) [reply src/raftstore/store/store.rs]: +1

**Pull request merged**